compiler/rustc_lint/src/mutable_transmutes.rs

```diff
@@ -0,0 +1,718 @@
+                            BuiltinMutablesTransmutes { from, to },
+                        );
+                        CheckOverlapping::No
+                    }
+                    _ => CheckOverlapping::Yes,
+                },
+                // For shared references, we need to see if they were transmuted from shared (not mutable!) references,
+                // and if there is an incorrectly transmuted `UnsafeCell` inside.
+                Mutability::Not => {
+                    let &ty::Ref(_, src_ty, Mutability::Not) = ty.kind() else {
+                        return CheckOverlapping::Yes;
+                    };
+                    debug!(?src_ty, dst_ty = ?dst_ref.ty, "found shared ref");
+
+                    check_unsafe_cells(
+                        cx,
+                        dst_ref.ty,
+                        src_ty,
+                        |src_pointee_fields_breadcrumbs, dst_pointee_fields_breadcrumbs| {
+                            let from = FieldsBreadcrumbs::format_unsafe_cell_for_diag(
+                                cx,
+                                src_fields_breadcrumbs,
+                                src,
+                                src_pointee_fields_breadcrumbs,
+                                src_ty,
+                            );
+                            let to = FieldsBreadcrumbs::format_unsafe_cell_for_diag(
+                                cx,
+                                &dst_ref.fields_breadcrumbs,
+                                dst,
+                                dst_pointee_fields_breadcrumbs,
+                                dst_ref.ty,
+                            );
+
+                            cx.emit_span_lint(
+                                UNSAFE_CELL_TRANSMUTES,
+                                expr.span,
+                                UnsafeCellTransmutes { from, to, orig_cast: None },
+                            );
+                        },
+                    );
+                    CheckOverlapping::No
+                }
+            }
+        },
+        &mut FieldsBreadcrumbs::default(),
+    );
+}
+
+fn get_transmute_from_to<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &Expr<'_>,
+) -> Option<(Ty<'tcx>, Ty<'tcx>)> {
+    let def = if let ExprKind::Path(ref qpath) = expr.kind {
+        cx.qpath_res(qpath, expr.hir_id)
+    } else {
+        return None;
+    };
+    if let Res::Def(DefKind::Fn, did) = def {
+        if !def_id_is_transmute(cx, did) {
+            return None;
+        }
+        let sig = cx.typeck_results().node_type(expr.hir_id).fn_sig(cx.tcx);
+        let from = sig.inputs().skip_binder()[0];
+        let to = sig.output().skip_binder();
+        return Some((from, to));
+    }
+    None
+}
+
+fn def_id_is_transmute(cx: &LateContext<'_>, def_id: DefId) -> bool {
+    cx.tcx.is_intrinsic(def_id, sym::transmute)
+}
+
+/// Checks for transmutes via `&*(reference as *const _ as *const _)` and lints.
```

Comment:
That's a cast, not a transmute...

Reply:
For the reasons described [here](https://github.com/rust-lang/rust/pull/128351#issuecomment-2530557427), I would actually call this a transmute.  It's using a cast to effect a transmutation.